internal/authz/authz.go

```diff
@@ -0,0 +1,327 @@
+		return CanCallReadAuthorizationModels, nil
+	case Read:
+		return CanCallRead, nil
+	case Write:
+		return CanCallWrite, nil
+	case ListObjects, StreamedListObjects:
+		return CanCallListObjects, nil
+	case Check:
+		return CanCallCheck, nil
+	case ListUsers:
+		return CanCallListUsers, nil
+	case WriteAssertions:
+		return CanCallWriteAssertions, nil
+	case ReadAssertions:
+		return CanCallReadAssertions, nil
+	case WriteAuthorizationModel:
+		return CanCallWriteAuthorizationModels, nil
+	case CreateStore:
+		return CanCallCreateStore, nil
+	case GetStore:
+		return CanCallGetStore, nil
+	case DeleteStore:
+		return CanCallDeleteStore, nil
+	case Expand:
+		return CanCallExpand, nil
+	case ReadChanges:
+		return CanCallReadChanges, nil
+	default:
+		return "", AuthorizationError{Err: ErrUnknownAPIMethod}.Err
+	}
+}
+
+// Authorize checks if the user has access to the resource.
+func (a *Authorizer) Authorize(ctx context.Context, storeID, apiMethod string, modules ...string) error {
+	claims, err := checkAuthClaims(ctx)
+	if err != nil {
+		return err
+	}
+
+	relation, err := a.getRelation(apiMethod)
+	if err != nil {
+		return err
+	}
+
+	// Check if there is top-level authorization first, before checking modules
+	err = a.individualAuthorize(ctx, claims.ClientID, relation, StoreIDType(storeID).String(), &openfgav1.ContextualTupleKeys{})
+	if err == nil {
+		return nil
+	}
+
+	if len(modules) > 0 {
+		return a.moduleAuthorize(ctx, claims.ClientID, relation, storeID, modules)
+	}
+	// If there are no modules to check, return the top-level authorization error
+	return err
+}
+
+// AuthorizeCreateStore checks if the user has access to create a store.
+func (a *Authorizer) AuthorizeCreateStore(ctx context.Context) error {
+	claims, err := checkAuthClaims(ctx)
+	if err != nil {
+		return err
+	}
+
+	relation, err := a.getRelation(CreateStore)
+	if err != nil {
+		return err
+	}
+
+	return a.individualAuthorize(ctx, claims.ClientID, relation, System, &openfgav1.ContextualTupleKeys{})
+}
+
+// GetModulesForWriteRequest returns the modules that should be checked for the write request.
+// If we encounter a type with no attached module, we should break and return no modules so that the authz check will be against the store
+// Otherwise we return a list of unique modules encountered so that FGA on FGA can check them after.
+func (a *Authorizer) GetModulesForWriteRequest(req *openfgav1.WriteRequest, typesys *typesystem.TypeSystem) ([]string, error) {
+	modulesMap := make(map[string]struct{})
+
+	tuples := make([]TupleKeyInterface, len(req.GetWrites().GetTupleKeys())+len(req.GetDeletes().GetTupleKeys()))
+	var index int
+	for _, tuple := range req.GetWrites().GetTupleKeys() {
+		tuples[index] = tuple
+		index++
+	}
+	index = 0
+	for _, tuple := range req.GetDeletes().GetTupleKeys() {
+		tuples[index] = tuple
+		index++
+	}
+
+	modulesMap, err := extractModulesFromTupleObjects(tuples, typesys, modulesMap)
```

Comment:
`modulesMap := make(map[string]struct{})`, why can't you crate this map within `extractModulesFromTupleObjects`?